### PR TITLE
Fix gcrs transforms for non geocentric frames

### PR DESCRIPTION
**Pull request by StuartLittlefair**
There was a bug when transforming to and from GCRS frames with non-geocentric coordinates (i.e those with non-zero values of ```obsgeoloc``` and ```obsgeovel```. This is because the ```erfa``` routine ```apcs13``` expects a position and velocity in GCRS coordinates.

Previously  ```obsgeoloc``` and ```obsgeovel``` were being passed naively, but these are defined in ITRS coordinates. I have added a helper routine ```get_gcrs_posvel``` to ```icrs_cirs_transforms``` to compute the proper pv array to supply to ```apcs13```.

### CODE REVIEW

**Comment by astrofrog** [comment]
@StuartLittlefair - can you do a rebase? It should get rid of the merge commits (you don't *need* to squash, I think merge commits get removed automatically)

File: astropy/coordinates/builtin_frames/icrs_cirs_transforms.py [hunk 1]
```diff
@@ -17,12 +17,56 @@
...
+        raise ValueError("frame must have obsgeoloc and obsgeovel attributes set")
+
+    # transform to GCRS. Just a rotation matrix so velocity should be fine?
+    geocentric_frame = GCRS(obstime=time)
+    pos_arr = obsgeoloc.transform_to(geocentric_frame).cartesian.xyz.to(u.m).value
+    vel_arr = obsgeovel.transform_to(geocentric_frame).cartesian.xyz.to(u.m/u.s).value
+
+    # now add rotational velocity of GCRS frame w.r.t to ITRS
+    # rotation of earth
+    V = u.Quantity([0, 0, 7.292115855306589e-5])*u.rad/u.s
```
Comment (astrofrog) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: Can this be defined as a constant, along with a reference for where it comes from?

**Comment by astrofrog** [comment]
@eteq will need to review this, but in the mean time @StuartLittlefair can you add a regression test, as well as a changelog entry in 1.1.3?

**Comment by StuartLittlefair** [comment]
Sorry @astrofrog; I don't know what you mean by a regression test...

**Comment by astrofrog** [comment]
@StuartLittlefair - ah sorry. What I mean is that you should add a test with an example that would have failed before this fix and that now works.

**Comment by astrofrog** [comment]
I think you could probably add the test in https://github.com/astropy/astropy/blob/ffc0a89b2c42fd440eb19bcb2f93db90cab3c98b/astropy/coordinates/tests/test_intermediate_transformations.py

File: astropy/coordinates/tests/test_intermediate_transformations.py [hunk 2]
```diff
@@ -128,6 +129,21 @@ def test_icrs_gcrs_dist_diff(gframe):
...
+    Test transformation to non-geocentric GCRS frames for nearby objects.
+    """
+    t = Time('1980-03-25 00:00')
+    gcrs_frame = GCRS(obstime=t,
+                      obsgeoloc=u.Quantity([-1463969.30185172,
+                                            -5166673.34223433,
+                                            3434985.71204565])*u.m)
+    icrs_coo = ICRS(ra=184.285*u.deg, dec=-1.8802*u.deg, distance=148101810.2*u.km)
+    gcrs_coo = icrs_coo.transform_to(gcrs_frame)
+
```
Comment (astrofrog) [inline astropy/coordinates/tests/test_intermediate_transformations.py]: Remove the spaces from line 143 (empty lines should not have spaces). This is causing a PEP8 failure on Travis.

**Comment by mhvk** [comment]
@StuartLittlefair - an annoying whitespace error... astropy/coordinates/tests/test_intermediate_transformations.py:143:1: W293 blank line contains whitespace

File: astropy/coordinates/builtin_frames/icrs_cirs_transforms.py [hunk 3]
```diff
@@ -17,12 +17,62 @@
...
 from .icrs import ICRS
+from .itrs import ITRS
 from .gcrs import GCRS
 from .cirs import CIRS
 from .hcrs import HCRS
 from .utils import get_jd12
 
 
+"""
+rotational velocity of Earth in radians per (UT1) second
```
Comment (mhvk) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: This confused me at first, since if it really was per UT1 seconds, it should just be 2 pi / 24 / 3600. But the rotational velocity given is actually per TAI second. Maybe rephrase the docstring
```
Rotational velocity of the Earth. In UT1 seconds, this would be 2 pi / (24 * 3600),
but we need the value in SI seconds.
```

File: astropy/coordinates/builtin_frames/icrs_cirs_transforms.py [hunk 4]
```diff
@@ -17,12 +17,62 @@
...
+    geocentric_frame = GCRS(obstime=time)
+    pos_arr = obsgeoloc.transform_to(geocentric_frame).cartesian.xyz.to(u.m).value
+    vel_arr = obsgeovel.transform_to(geocentric_frame).cartesian.xyz.to(u.m/u.s).value
+
+    # now add rotational velocity of GCRS frame w.r.t to ITRS
+    extra_rotation = np.cross(V_EARTH, np.rollaxis(pos_arr, 0, pos_arr.ndim))
+    vel_arr += np.rollaxis(extra_rotation, -1, 0)
+
+    pv = np.array([pos_arr, vel_arr])
+    if len(pv.shape) > 2:
```
Comment (mhvk) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: Are you sure this is needed? This routine replaces a line that just does the `np.array([pos, vel])`

**Comment by mhvk** [comment]
See some further small comments. I'm convinced enough that this is right to merge it with those in place, but ideally @eteq has a look too.

File: astropy/coordinates/builtin_frames/icrs_cirs_transforms.py (hunk 4 above)
Comment (StuartLittlefair) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: I believe it's necessary because the ```frame``` might have non-scalar times, whereas the frame's ```obsgeoloc``` etc were always scalar.

Comment (mhvk) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: If so, maybe explicitly write `if time.ndim > 1:`?
Or add a comment here... (can be commit with [skip ci])

File: CHANGES.rst [hunk 5]
```diff
@@ -503,6 +503,11 @@ Bug Fixes
 
 - ``astropy.coordinates``
 
+  - Transformations between ``GCRS`` frames and other frames now properly handles
+    frames with non-zero ```obsgeoloc```. Previously this was in slight error for
+    nearby objects, leading to position errors of a few arcseconds for solar system
+    planets, to a degree for the Moon.
```
Comment (eteq) [inline CHANGES.rst]: Add ``[#4894]`` to the end of this just so interested parties get a link to this issue.

**Comment by eteq** [comment]
@StuartLittlefair - Sorry, I'm confused by this.  Why do you say ``obsgeoloc`` and ``obsgeovel`` are in ITRS?  The documentation for them says that they are "oriented the same as BCRS/ICRS", which means they are *not* ITRS, right?  Or am I missing something?

**Comment by StuartLittlefair** [comment]
@eteq poo. the documentation does say this, I had missed that.
 
This PR came about because @bmorris was initialising the ```obsgeoloc``` attribute of a frame with the ```geocentric``` attribute of an ```EarthLocation``` and I didn't realise this was incorrect. Perhaps the correct place for the function I've written here is a ```get_gcrs_posvel``` attribute of an ```EarthLocation```?

This raises a slight issue:

The documentation also says ```obsgeoloc``` and ```obsgeovel``` should be 3-vectors, which is not really possible in the use case of a single observatory on Earth and non-scalar times, in which case the GCRS position/velocity is time dependent. Would it be an issue if they had higher dimensionality?

**Comment by eteq** [comment]
@StuartLittlefair - yep, your solution of putting `get_gcrs_posvel` in `EarthLocation` sounds perfect - hopefully you can just copy-and-paste this there. (presumably as part of #4890, or as a separate quick/small PR)

As for "3-vectors", I'm 95% sure that's a subtle lie: it's actually that the *trailing* (or maybe leading?) dimension has to be 3, but otherwise it can be whatever dimensionality is required.  "3-vector" is just there to indicate they are cartesian x/y/z in the same frame as GCRS, rather than spherical or whatever.  Feel free to also clear up those docs (ideally once you've figured out if it's trailing or leading... easiest way to find out is try one or the other and see which raises exceptions...)

**Comment by eteq** [comment]
Oh, and I'm going to kill the tests here just because I'm pretty sure you'll want to close this given the discussion above.  But I can re-start them if we end up thinking this is needed after all.

**Comment by StuartLittlefair** [comment]
No - I'm pretty sure this is not needed. I'm going to add the changes to EarthLocation in #4890.

**Comment by eteq** [comment]
OK, great.  Once you've decided that for sure, @StuartLittlefair, feel free to just close this.
